performance.Rmd

```diff
@@ -0,0 +1,184 @@
+the code and it is executed line by line.
+
+Most of the time, this kind of execution is perfectly fine. With datasets of a
+reasonable size (let's say up to a few hundreds of thousands of observations and
+a few dozens columns), we don't really have to worry about whether our code is
+optimized to save memory and time.
+
+However, when we start dealing with much larger datasets, the order in which
+functions are applied is extremely important. Indeed, some functions are much
+more memory intensive than others.
+
+For instance, let's say we have a dataset with 1M observations containing
+country names and a few numeric columns. We would like to keep just a few of
+these countries and sort them alphabetically. Here, we have two operations:
+filtering and sorting. Sorting is much harder to do than filtering. To filter
+data, we simply check whether each row fills some conditions, but to sort data
+we have to compare rows between them and rearrange them as we go. If we don't
+take this into account, sorting data before filtering it can slow down our
+pipeline significantly.
+
+```{r}
+countries = c("France", "Germany", "United Kingdom", "Japan", "Columbia", 
+               "South Korea", "Vietnam", "South Africa", "Senegal", "Iran")
+
+set.seed(123)
+test = data.frame(
+  country = sample(countries, 1e6, TRUE),
+  x = sample(1:100, 1e6, TRUE),
+  y = sample(1:1000, 1e6, TRUE)
+)
+
+bench::mark(
+  sort_filter = {
+    tmp = test[order(test$country), ]
+    subset(tmp, country %in% c("United Kingdom", "Japan", "Vietnam"))
+  },
+  filter_sort = {
+    tmp = subset(test, country %in% c("United Kingdom", "Japan", "Vietnam"))
+    tmp[order(tmp$country), ]
+  }
+)
+```
+
+
+## How does `polars` help?
+
+We have seen that the order in which functions are applied matters a lot. But it
+already takes a long time to deal with large data, we're not going to spend even
+more time trying to optimize our pipeline, right?
+
+This is where *lazy* execution comes into play. The idea is that we write our
+code as usual, but this time, we won't apply it directly on a dataset but on a
+lazy dataset, i.e a dataset that is not loaded in memory yet (in `polars` terms,
+these are `DataFrame`s and `LazyFrame`s). Once our code is ready, we
+call `collect()` at the end of the pipeline. Before executing our code, `polars`
+will internally check whether it can be optimized, for example by reordering
+some operations.
+
+Let's re-use the example above but this time with `polars` syntax and 10M
+observations. It doesn't make sense to convert a dataset that we already have in
+memory to a `LazyFrame` so first we have to export it as CSV and then we can
+read it lazily with `lazy_csv_reader()`:
+
+```{r echo = FALSE, eval = !exists("data/test.csv")}
+set.seed(123)
+test = data.frame(
+  country = sample(countries, 1e7, TRUE),
+  x = sample(1:100, 1e7, TRUE),
+  y = sample(1:1000, 1e7, TRUE)
+)
+readr::write_csv(test, "data/test.csv")
```

Comment:
maybe readr should be added to suggested

Reply:
Probably should just use the base `write.csv` instead. No reason to load a new dependency just for writing a CSV...

Reply:
sure, I just used `readr` because it much faster than `write.csv` but it's not necessary, I removed it